src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php

```diff
@@ -867,4 +895,454 @@ public static function arithmeticOperation(
+            } elseif ($right_type_part->isNegative()) {
+                $new_result_type = Type::getFloat();
+            } elseif ($right_type_part->min_bound === 0 && $right_type_part->max_bound === 0) {
+                $new_result_type = Type::getInt(true, 1);
+            } else {
+                //$left_type_part may be a mix of positive, negative and 0
+                $new_result_type = new Type\Union([new TInt(), new TFloat()]);
+            }
+        }
+
+        if (!$result_type) {
+            $result_type = $new_result_type;
+        } else {
+            $result_type = Type::combineUnionTypes($new_result_type, $result_type);
+        }
+    }
+
+    private static function analyzeModBetweenIntRange(
+        TIntRange $right_type_part,
+        TIntRange $left_type_part,
+        ?Type\Union &$result_type
+    ): void {
+        //result of Mod is not directly dependant on the bounds of the range
+        if ($right_type_part->min_bound !== null && $right_type_part->min_bound === $right_type_part->max_bound) {
+            //if the second operand is a literal, we can be pretty detailed
+            if ($right_type_part->min_bound === 0) {
+                $new_result_type = Type::getEmpty();
+            } else {
+                if ($left_type_part->isPositiveOrZero()) {
+                    if ($right_type_part->isPositive()) {
+                        $max = $right_type_part->min_bound - 1;
+                        $new_result_type = new Type\Union([new TIntRange(0, $max)]);
+                    } else {
+                        $max = $right_type_part->min_bound + 1;
+                        $new_result_type = new Type\Union([new TIntRange($max, 0)]);
+                    }
+                } elseif ($left_type_part->isNegativeOrZero()) {
+                    if ($right_type_part->isPositive()) {
+                        $max = $right_type_part->min_bound - 1;
+                        $new_result_type = new Type\Union([new TIntRange(-$max, 0)]);
+                    } else {
+                        $max = $right_type_part->min_bound + 1;
+                        $new_result_type = new Type\Union([new TIntRange(-$max, 0)]);
+                    }
+                } else {
+                    if ($right_type_part->isPositive()) {
+                        $max = $right_type_part->min_bound - 1;
+                    } else {
+                        $max = -$right_type_part->min_bound - 1;
+                    }
+                    $new_result_type = new Type\Union([new TIntRange(-$max, $max)]);
+                }
+            }
+        } elseif ($right_type_part->isPositive()) {
+            if ($left_type_part->isPositiveOrZero()) {
+                $new_result_type = new Type\Union([new TIntRange(0, null)]);
+            } elseif ($left_type_part->isNegativeOrZero()) {
+                $new_result_type = new Type\Union([new TIntRange(null, 0)]);
+            } else {
```

Comment:
This seemed pretty complex, so I didn't try. If we fall into this else, this means we have something like
`int<57, 95> % int<min, 18>`. I guess I could put a PositiveInt, because a negative in modulo seem to have the same effective as if it was positive so it would result in a positive integer

Reply:
I meant the whole `if..elseif..else` block.

Reply:
I guess I could make a case for degenerate right_type_part here
Now that I think about it, if we have a PositiveOrZero range in left, we can pretty much transform it into a int<0, left_max-1>